clients/src/main/java/org/apache/kafka/clients/admin/DescribeClusterResults.java

```diff
@@ -28,16 +28,38 @@
  */
 @InterfaceStability.Unstable
 public class DescribeClusterResults {
-    private final KafkaFuture<Collection<Node>> future;
+    private final KafkaFuture<Collection<Node>> nodesFuture;
+    private final KafkaFuture<Node> controllerFuture;
+    private final KafkaFuture<String> clusterIdFuture;
 
-    DescribeClusterResults(KafkaFuture<Collection<Node>> future) {
-        this.future = future;
+    DescribeClusterResults(KafkaFuture<Collection<Node>> nodesFuture,
+                           KafkaFuture<Node> controllerFuture,
+                           KafkaFuture<String> clusterIdFuture) {
+        this.nodesFuture = nodesFuture;
+        this.controllerFuture = controllerFuture;
+        this.clusterIdFuture = clusterIdFuture;
     }
 
     /**
      * Returns a future which yields a collection of nodes.
      */
     public KafkaFuture<Collection<Node>> nodes() {
-        return future;
+        return nodesFuture;
+    }
+
+    /**
+     * Returns a future which yields the current controller id.
+     * Note that this may yield null, if the controller ID is not yet known.
+     */
+    public KafkaFuture<Node> controller() {
+        return controllerFuture;
+    }
+
+    /**
+     * Returns a future which yields the current cluster Id.
+     * Note that this may yield null, if the cluster version is too old.
+     */
+    public KafkaFuture<String> clusterIdFuture() {
```

Comment:
@cmccabe to clarify my point: if you have 3 futures and you want to use all of them, you need to a way to combine the 3 futures into 1 and handle the fact that each may fail separately.

So, imagine that someone wants to use these 3 fields somewhere in their code after handling errors. I'll use Scala Future terminology next. The typical way to do these things in a non-blocking API is to map/flatMap over the future and pass the unwrapped value to other code. The code, in turn, may return a `Future` if it needs to perform some computation that is expensive, involves IO, etc or simply return a normal value. In the former case, `flatMap` is needed why in the latter `map` will do. If you have 3 `Future`s instead of `1`, you need to do 3 nested `flatMap` calls and pass the individual values to a method or create a class containing these values before passing them around. So, in general, it's easier if this is done for you.

We can always introduce separate futures in the future if there's a case for handling errors separately. I can't think of a good reason to do that here though.